Fixed #25052 -- Added password validation to UserCreationForm.

timgraham opened the pull request:
https://code.djangoproject.com/ticket/25052

timgraham commented on django/contrib/auth/forms.py:
```diff
@@ -86,11 +86,15 @@ def clean_password2(self):
-        user.set_password(self.cleaned_data["password1"])
+        password = self.cleaned_data['password1']
+        user.set_password(password)
+        password_validation.password_changed(password, self.instance)
```
This positioning is copied from the other forms, but I think it's wrong here and there. If you use `form.save(commit=False)`, you'll get the `password_changed()` functions called twice. If you have a function that prevents password reuse, that'll be a problem. Also these lines appear to be untested.

timgraham commented:
@carljm @erikr - do these changes look okay?

carljm commented:
Looks mostly ok to me. I don't have a strong intuition that `commit=False` shouldn't call the `password_changed` hook; did someone actually raise that as an issue? ISTM if we make that change, we also need to add docs mentioning the behavior. 

Personally I would be inclined to call `password_changed` regardless. The intended use cases for `commit=False` still involve saving the object eventually.

timgraham commented:
Sorry, my explanation comment above got hidden: "If you use `form.save(commit=False)`, you'll get the `password_changed()` functions called twice. If you have a function that prevents password reuse, that'll be a problem." Sure, I'll add docs unless you see a problem with this.

timgraham commented:
The docs say, "You can also implement ``password_changed(password, user=None``), which will
be called after a successful password change". It makes sense to me that it shouldn't be called until after the user is actually saved, since something could go wrong before that. Do you think the docs need to be clarified? If the current behavior is more desirable, then I don't see why `password_validation.password_changed()` shouldn't be put in the `User.set_password()` method.

carljm commented:
The doc addition that I think is needed is to warn users that if you use `.save(commit=False)` on any of these forms, you are responsible for calling `validation.password_changed(...)` yourself after saving the user. I highly doubt that anyone would think to do so otherwise. Unless I'm missing something, I think this doc addition is probably needed in the docs for every form which has this behavior.

I'm not sure I understand the context you refer to above in which `password_changed(...)` might be called twice on the same password change.

The reason I was leaning towards just calling it always (and sure, it could just be moved directly into `set_password` in that case I guess) is that I think if someone is using `form.save(commit=False)`, forgetting to call `password_changed` themselves is a more likely source of bugs than "something goes wrong before actually saving the user." On the other hand, forgetting to call it yourself is easily fixed, whereas if the user save is actually conditional, you can't easily undo a `password_changed` call that was already made. So I guess your approach is preferable for that reason. But I definitely think the doc additions are needed.

timgraham commented:
Oh, that's a good point and something I missed. I had another idea to do something like an addition in the `user.save()` method that would check a "password dirty" flag that `set_password()` would set and trigger `password_changed` after a successful save. I had a concern that custom user models would have to obey this contract, but maybe it's an acceptable tradeoff. Worth thinking about some more?

carljm commented:
Hmm, that is an interesting idea. `set_password()` is implemented on `AbstractBaseUser`, the most stripped-down abstract base user class we provide. I think it's quite likely that virtually all custom user models that use Django's password authentication system probably inherit `AbstractBaseUser` already. (I assume you'd add the `save()` method there too). A note could be added in the custom auth model documentation that if you don't inherit `AbstractBaseUser` but want to use Django password validation, you have to take care of calling `password_changed` at the appropriate time yourself. That seems reasonable; maybe the most reasonable option of all the ones we've discussed.

That seems pretty good to me. I'm curious whether @erikr has an opinion here.

timgraham commented:
I implemented the alternate approach. If it looks fine, I guess some documentation updates could be added too.

carljm commented:
Seems ok to me. It's a little weird that most password validation stuff is at the form layer (where _validation_ should be), but then the `password_changed` call happens at the model layer. And I dunno if there'd be any concern about keeping the raw password around in memory longer; I suppose it might raise the risk that it could be accidentally dumped as part of some kind of debugging output?

On the other hand, I think this approach gives the best behavior: `password_changed` is called automatically for you in all cases (unless you're not using `AbstractBaseUser`), but only when the password change has actually been successfully finalized.

timgraham commented:
I think this is ready for a final review. Since last time, I added a paragraph to the docs and also fixed the problem of a password hash upgrade triggering `password_changed()`.

carljm commented on django/contrib/auth/base_user.py:
```diff
@@ -60,9 +61,20 @@ def get_username(self):
+        super(AbstractBaseUser, self).save(*args, **kwargs)
+        if self._password is not None:
+            password_validation.password_changed(self._password, self)
+
```
I think we should clear `self._password` here to a) minimize the lifespan of the raw password in memory and b) avoid double-calling `password_changed` in case of multiple subsequent saves of the same user instance.

carljm commented:
LGTM modulo the above comment.

erikr commented:
Apologies, I didn't have time to look into this now. Fortunately, I completely concur with this approach. This seems the least error prone and overall a clean design :)
